### PR TITLE
[NTOS:MM] Fix ntoskrnl/mm/ARM3/virtual.c causing intermittent BSOD 0x1a

### PR DESCRIPTION
**Pull request by Doug-Lyons**
This was introduced by commit https://github.com/reactos/reactos/commit/c7e09061ca
So, this is basically just a revert of this commit. I did move a block of the reverted code from miarm.h to virtual.c.

## Purpose

_Fix intermittent BSOD 0x0000001a in mm/ARM3/virtual.c._

JIRA issue: [CORE-18190](https://jira.reactos.org/browse/CORE-18190)
JIRA issue: [CORE-18818](https://jira.reactos.org/browse/CORE-18818)
JIRA issue: [CORE-19253](https://jira.reactos.org/browse/CORE-19253)
and probably others.

## Proposed changes

_Mostly revert all of the referenced commit in two out of three files._
_One file is left unreverted (page.c). Combine changes in other two files into new virtual.c changes._
_There was a code block regarding ```__MI_PAGING_LEVELS``` in 'ntoskrnl\mm\ARM3\miarm.h'._
_These lines are what I moved into virtual.c now to keep all this related code together._
_I left a ```#if 0``` block that contains the problematic code in place but disabled._
_This was to allow a future fix of the BSOD 0x1a caused by this code to be fixed more easily._

_Summary: Restore and use MiQueryPageTableReferences._

Testbot results:
PR#66633 refs/pull/6633/merge on top of 0.4.15-dev-7786-gc5e6456

VBox: https://reactos.org/testman/compare.php?ids=94065,94069 LGTM
KVM:  https://reactos.org/testman/compare.php?ids=94064,94070 LGTM

### CODE REVIEW

File: ntoskrnl/mm/ARM3/virtual.c
Comment (SergeGautherie) [inline ntoskrnl/mm/ARM3/virtual.c:22]: `#if` could be inside the function.
Do we want to add parentheses?

File: ntoskrnl/mm/ARM3/virtual.c
Comment (SergeGautherie) [inline ntoskrnl/mm/ARM3/virtual.c:746]: What is the purpose of keeping this block?

**Review by SergeGautherie — COMMENTED** [review]
> Mostly revert all of the reference commit in two out of three files.

Something looks inconsistent : only virtual.c is modified.

--

> [NTOS:MM] Fix ntoskrnl/mm/ARM3/virtual.c causing BSOD 0x1a introduced  y commit https://github.com/reactos/reactos/commit/c7e09061ca

Please, tell what the functional change is, instead.

**Comment by julenuri** [comment]
My tests:
Blender x86 and x64 from [CORE-18720](https://jira.reactos.org/browse/CORE-18720) are working without crashing.
API Mon x86 and x64 is not crashing every closing app. [CORE-18818](https://jira.reactos.org/browse/CORE-18818) really important for internal investigations!!!
Dolphin emulator, VBA-M and PCSX2 can stop the emulation of a game without crashing.
Game Dev Tycoon is neither failing [CORE-19253](https://jira.reactos.org/browse/CORE-19253) , I have been able to play the full demo without crashing:
VirtualBox_ReactOS_17_03_2024_18_36_32
Chrome didn't fail for me the time I have been using it. That is nice, in both arch, x86 and x64 [CORE-19279](https://jira.reactos.org/browse/CORE-19279)
Spotify now closes succesfully and have a tiny repainting bug of the window borders after closing it. [CORE-18190](https://jira.reactos.org/browse/CORE-18190)

Not tested the RVIC bible software, the Pelles C and the Yandex seems to not fail, but I haven't been able to make it install fully.

File: ntoskrnl/mm/ARM3/virtual.c
Comment (Doug-Lyons) [reply ntoskrnl/mm/ARM3/virtual.c:746]: This was the change that caused the problem. I wanted to leave it as a reminder in hopes that a developer could actually make it work without the BSOD's. Then the code could be simplified in accordance with the original commit.

File: ntoskrnl/mm/ARM3/virtual.c
Comment (Doug-Lyons) [reply ntoskrnl/mm/ARM3/virtual.c:22]: This was a copy of the original deleted code as it existed before the problem commit. It was originally in 'ntoskrnl/mm/ARM3/miarm.h' and  I moved it here to be able to keep all of this code in a single file. This pattern using the ```__MI_PAGAING_LEVELS``` appears in this same pattern in other places in the code, so I feel that it should not be modified here. Thanks.

**Comment by Doug-Lyons** [comment]
> Combine changes in other two files into new virtual.c changes.

> Something looks inconsistent : only virtual.c is modified.

There was minimal code regarding ```__MI_PAGING_LEVELS``` in 'ntoskrnl\mm\ARM3\miarm.h' that I moved into virtual.c here.

**Review by JoachimHenze — APPROVED** [review]

**Review by DarkFire01 — APPROVED** [review]
My knowledge on Mm is WIP, I’ve been studying it but admittedly I am not super confident in my comprehension.

having said that looking at what this is doing I see no clear down side to accepting this revert/change. I guess wait for 2 more core devs before merge :D

**Review by julenuri — APPROVED** [review]
For me, to be merged. I only see positive behavior.

**Review by tkreuzer — APPROVED** [review]
I had another look to see, if I could find what's really wrong here, but it didn't lead to much. I think the revert is fine. I will bookmark it, because I liked the original code, and at som point I might have the time and dedication to redo this properly. For now I think, it's go.

**Comment by Doug-Lyons** [comment]
Thank you very much, @tkreuzer for the commit and also @JoachimHenze, @julenuri, and @DarkFire01 for your approvals and input on this. And thanks @SergeGautherie for your review comments.

**Pull request merged**
